src/EventStore.Core/Services/Archiver/ArchiverService.cs

```diff
@@ -2,54 +2,192 @@
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using EventStore.Core.Bus;
+using EventStore.Core.Data;
+using EventStore.Core.Messages;
 using EventStore.Core.Services.Archiver.Storage;
-using EventStore.Plugins;
-using EventStore.Plugins.Licensing;
-using Microsoft.AspNetCore.Builder;
-using Microsoft.Extensions.Configuration;
-using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.Logging;
-using ILogger = Serilog.ILogger;
+using EventStore.Core.Services.Archiver.Storage.Exceptions;
+using Serilog;
 
 namespace EventStore.Core.Services.Archiver;
 
-public class ArchiverService : IPlugableComponent {
-	protected static readonly ILogger Log = Serilog.Log.ForContext<ArchiverService>();
+public class ArchiverService :
+	IHandle<SystemMessage.ChunkLoaded>,
+	IHandle<SystemMessage.ChunkCompleted>,
+	IHandle<SystemMessage.ChunkSwitched>,
+	IHandle<ReplicationTrackingMessage.ReplicatedTo>,
+	IHandle<SystemMessage.SystemStart>,
+	IHandle<SystemMessage.BecomeShuttingDown>
+{
+	private static readonly ILogger Log = Serilog.Log.ForContext<ArchiverService>();
 
-	public string Name => "Archiver";
+	private readonly ISubscriber _mainBus;
+	private readonly IArchiveStorage _archiveStorage;
+	private readonly Queue<ChunkInfo> _uncommittedChunks;
+	private readonly ConcurrentDictionary<string, ChunkInfo> _existingChunks;
+	private readonly CancellationTokenSource _cts;
+	private readonly Channel<Commands.ArchiveChunk> _archiveChunkCommands;
 
-	public string DiagnosticsName => Name;
+	private readonly TimeSpan RetryInterval = TimeSpan.FromMinutes(1);
+	private long _replicationPosition;
 
-	public KeyValuePair<string, object>[] DiagnosticsTags => [];
+	public ArchiverService(ISubscriber mainBus, IArchiveStorage archiveStorage) {
+		_mainBus = mainBus;
+		_archiveStorage = archiveStorage;
 
-	public string Version => "0.0.1";
+		_uncommittedChunks = new();
+		_existingChunks = new();
+		_cts = new();
+		_archiveChunkCommands = Channel.CreateUnboundedPrioritized(
+			new UnboundedPrioritizedChannelOptions<Commands.ArchiveChunk> {
+			SingleWriter = true,
+			SingleReader = true,
+			Comparer = new ChunkPrioritizer()
+		});
 
-	public bool Enabled => true;
+		Subscribe();
+		Start();
+	}
+
+	private void Subscribe() {
+		_mainBus.Subscribe<SystemMessage.ChunkLoaded>(this);
+		_mainBus.Subscribe<SystemMessage.ChunkSwitched>(this);
+		_mainBus.Subscribe<SystemMessage.ChunkCompleted>(this);
+		_mainBus.Subscribe<ReplicationTrackingMessage.ReplicatedTo>(this);
+		_mainBus.Subscribe<SystemMessage.SystemStart>(this);
+		_mainBus.Subscribe<SystemMessage.BecomeShuttingDown>(this);
+	}
+
+	private void Start() {
+		Task.Factory.StartNew(
+			() => ArchiveChunks(_cts.Token),
+			_cts.Token,
+			TaskCreationOptions.LongRunning,
```

Comment:
LongRunning is suitable when the task is going to use the thread that its scheduled on for a long time (it's a hint to the scheduler that another thread might be handy)   but that isn't the case here. we can just use task.run

Reply:
fixed, thanks for the clarification